src/env/hedged/mod.rs

```diff
@@ -0,0 +1,671 @@
+            path1,
+            path2,
+            sender,
+            seqno1,
+            seqno2,
+            thread1: Some(thread1),
+            thread2: Some(thread2),
+        }
+    }
+
+    pub fn state(&self) -> State {
+        self.sender.state()
+    }
+
+    async fn wait_handle(&self, task1: Task, task2: Task) -> IoResult<HedgedHandle> {
+        let (cb1, mut f1) = paired_future_callback();
+        let (cb2, mut f2) = paired_future_callback();
+        self.sender.send(task1.clone(), task2.clone(), cb1, cb2);
+
+        let resolve = |res: TaskRes| -> (LogFd, bool) {
+            match res {
+                TaskRes::Create { fd, is_for_rewrite } => (fd, is_for_rewrite),
+                TaskRes::Open { fd, is_for_rewrite } => (fd, is_for_rewrite),
+                _ => unreachable!(),
+            }
+        };
+        select! {
+            res1 = f1 => res1.unwrap().map(|res| {
+                let (fd, is_for_rewrite) = resolve(res);
+                HedgedHandle::new(
+                    self.base.clone(),
+                    is_for_rewrite,
+                    self.sender.clone(),
+                    FutureHandle::new_owned(fd),
+                    FutureHandle::new(f2, task2),
+                    self.seqno1.clone(),
+                    self.seqno2.clone(),
+                )
+            }),
+            res2 = f2 => res2.unwrap().map(|res| {
+                let (fd, is_for_rewrite) = resolve(res);
+                HedgedHandle::new(
+                    self.base.clone(),
+                    is_for_rewrite,
+                    self.sender.clone(),
+                    FutureHandle::new(f1, task1),
+                    FutureHandle::new_owned(fd) ,
+                    self.seqno1.clone(),
+                    self.seqno2.clone(),
+                )
+            }),
+        }
+    }
+
+    async fn wait(&self, task1: Task, task2: Task) -> IoResult<()> {
+        let (cb1, mut f1) = paired_future_callback();
+        let (cb2, mut f2) = paired_future_callback();
+        self.sender.send(task1, task2, cb1, cb2);
+
+        select! {
+            res1 = f1 => res1.unwrap().map(|_| ()),
+            res2 = f2 => res2.unwrap().map(|_| ()),
+        }
+    }
+}
+
+impl Drop for HedgedFileSystem {
+    fn drop(&mut self) {
+        block_on(self.wait(Task::Stop, Task::Stop)).unwrap();
+
+        let t1 = self.thread1.take().unwrap();
+        let t2 = self.thread2.take().unwrap();
+        let mut times = 0;
+        loop {
+            if t1.is_finished() && t2.is_finished() {
+                t1.join().unwrap();
+                t2.join().unwrap();
+                break;
+            }
+            times += 1;
+            if times > 100 {
+                // wait 1s
+                // one disk may be blocked for a long time,
+                // to avoid block shutdown process for a long time, do not join the threads
+                // here, only need at least to ensure one thread is exited
```

Comment:
HedgedFileSystem is dropped but its underlying thread may still live, which is not ideal. 

Maybe we should abort the pending tasks on slow disk and wait both threads exist.